Common/UnitRelations.json

```diff
@@ -0,0 +1,165 @@
+[
+  { "Operator": "*", "Result": "AmountOfSubstance.FromKilomoles", "Left": "MolarFlow.KilomolesPerSecond", "Right": "Duration.Seconds" },
+  { "Operator": "*", "Result": "Angle.FromRadians", "Left": "RotationalSpeed.RadiansPerSecond", "Right": "Duration.Seconds" },
+  { "Operator": "*", "Result": "Area.FromSquareMeters", "Left": "KinematicViscosity.SquareMetersPerSecond", "Right": "Duration.Seconds" },
+  { "Operator": "*", "Result": "Area.FromSquareMeters", "Left": "Length.Meters", "Right": "Length.Meters" },
+  { "Operator": "*", "Result": "DynamicViscosity.FromNewtonSecondsPerMeterSquared", "Left": "Density.KilogramsPerCubicMeter", "Right": "KinematicViscosity.SquareMetersPerSecond" },
+  { "Operator": "*", "Result": "ElectricCharge.FromAmpereHours", "Left": "ElectricCurrent.Amperes", "Right": "Duration.Hours" },
+  { "Operator": "*", "Result": "ElectricCurrent.FromAmperes", "Left": "ElectricCurrentGradient.AmperesPerSecond", "Right": "Duration.Seconds" },
+  { "Operator": "*", "Result": "ElectricPotential.FromVolts", "Left": "ElectricCurrent.Amperes", "Right": "ElectricResistance.Ohms" },
+  { "Operator": "*", "Result": "Energy.FromJoules", "Left": "ElectricPotential.Volts", "Right": "ElectricCharge.Coulombs" },
+  { "Operator": "*", "Result": "Energy.FromJoules", "Left": "EnergyDensity.JoulesPerCubicMeter", "Right": "Volume.CubicMeters" },
+  { "Operator": "*", "Result": "Energy.FromJoules", "Left": "Power.Watts", "Right": "Duration.Seconds" },
+  { "Operator": "*", "Result": "Energy.FromJoules", "Left": "SpecificEnergy.JoulesPerKilogram", "Right": "Mass.Kilograms" },
+  { "Operator": "*", "Result": "Energy.FromJoules", "Left": "TemperatureDelta.Kelvins", "Right": "Entropy.JoulesPerKelvin" },
+  { "Operator": "*", "Result": "Entropy.FromJoulesPerKelvin", "Left": "SpecificEntropy.JoulesPerKilogramKelvin", "Right": "Mass.Kilograms" },
+  { "Operator": "*", "Result": "Force.FromNewtons", "Left": "ForceChangeRate.NewtonsPerSecond", "Right": "Duration.Seconds" },
+  { "Operator": "*", "Result": "Force.FromNewtons", "Left": "ForcePerLength.NewtonsPerMeter", "Right": "Length.Meters" },
+  { "Operator": "*", "Result": "Force.FromNewtons", "Left": "Mass.Kilograms", "Right": "Acceleration.MetersPerSecondSquared" },
+  { "Operator": "*", "Result": "Force.FromNewtons", "Left": "Pressure.Pascals", "Right": "Area.SquareMeters" },
```

Comment:
I think this is a sound design.

### 1. How about making it more compact
It could make it easier to read and maintain. It also resembles code using these operators.

This structure should be possible to parse, and CodeGen has access to the the plural form of units to substitute in the generated code.

```json
[ 
"ElectricCurrent.Ampere * Duration.Hour = ElectricCharge.AmpereHour",
"ElectricCurrent.Ampere * ElectricResistance.Ohm = ElectricPotential.Volt",

"Power.Watt / ElectricPotential.Volt = ElectricCurrent.Ampere",

"1 / Length.Meter = ReciprocalLength.InverseMeter"
]
```

### 2. Document sort order
One weakness with this approach, and the old approach too, is that it is not easy to see if a particular relation is defined. In particular for multiplication, where only one of two permutations of left/right is defined.

I propose a sorting the the document alphabetically on: 
- Left side argument first, then by right side argument

This reads better when following the string-based schema above, or changing the JSON to move "Result" last.

It's going to be a pain to enforce this for contributors, but I think we should at least define a standard layout/order for the document that we can maybe even have CodeGen auto-sort for us.


Let me know what you think.

Reply:
Just realized, since multiplication and possibly addition only defines one of two permutations, `sorting on left side argument first` is ambiguous on what quantity to choose as the left side.

### Option 2.1
Pick the left/right side by alphabetical order also, so `ElectricCurrent.Ampere * Duration.Hour = xxx` would instead be `Duration.Hour * ElectricCurrent.Ampere = xxx`.

### Option 2.2
Or we could forego this problem altogether and move the `result` first and sort on that, similar to your schema. 

Either like this:
- Result type (alphabetically)
- Then operation order (`* / + -`)


```json
[
"ElectricCharge.AmpereHour = ElectricCurrent.Ampere * Duration.Hour",
"ElectricCurrent.Ampere = Power.Watt / ElectricPotential.Volt",
"ElectricPotential.Volt = ElectricCurrent.Ampere * ElectricResistance.Ohm",
"ReciprocalLength.InverseMeter = 1 / Length.Meter"
]
```

Or like this:
- Operation order (`* / + -`)
- Then result type (alphabetically)

```json
[ 
// Multiplication
"ElectricCharge.AmpereHour = ElectricCurrent.Ampere * Duration.Hour",
"ElectricPotential.Volt = ElectricCurrent.Ampere * ElectricResistance.Ohm",

// Division, including inverse
"ReciprocalLength.InverseMeter = 1 / Length.Meter",
"ElectricCurrent.Ampere = Power.Watt / ElectricPotential.Volt",
]
```

### Option 2.3
Duplicate all two-way methods, and possibly use CodeGen to enforce this by first scanning and correcting the document before parsing it.

Reply:
Thanks, awesome feedback. The schema was the thing I was least happy with and I like your options better.

I've reworked this like option 2.2, but initially just sorted alphabetically on the whole string. Also added:

- Auto-sort and write to disk on codegen
- Validation of relations based on unit definitions

Reply:
> Another idea: generate division operators based on multiplication

Interesting idea, my gut tells me this may be a notch too much magic, since it is not clear whether a division operator implementation comes from a multiplication definition or an explicit division definition. Also, there are some dragons with some of the quantities, and maybe some division operations either don't make sense or are not desirable.

Then again, we can just try and see how it works out. I'm all for experimenting a bit on this 👍

Reply:
> alphabetically on the whole string

That works too, simpler even.

Reply:
I did some experimenting with generating division operators based on multiplication. I do agree it might be a bit magic, but I managed to avoid some confusion by removing all explicit division definitions 😄 
There are indeed some dragons so I had to pull a few tricks, but in the end it seems to work. See #1354

Reply:
Cool! Let's continue the division experiment separately after merging this first, I think the two changes can be isolated.

Is there anything else that remains on this PR now, or is it ready for the final review?

Reply:
While it stays backwards compatible, the API surface grows quite a bit, and the units get generic math interfaces for the first time. It's also a conceptual change for contributors. Wouldn't we want to target v6?

Reply:
No I think it's fine. 
We're just adding stuff and I don't see anything too experimental here that needs to mature its design before release.

Reply:
Allright, I pushed some changes to clear up some code, add some comments and sort the operators after inferring new ones. Ready for final review 👍